### ISSUE TITLE
How can I enable IPIP RSS on my L7 server?(the second time

### ISSUE THREAD

**Issue by takehaya** [facebookincubator/katran#112]

**Comment by udippant** [comment]
Can you paste the output of `ethtool --show-offload eth0` ?

**Comment by takehaya** [comment]
ofcourse :)
```
# ethtool --show-offload ens4f0                                                               
Features for ens4f0:                                                                                                   
rx-checksumming: off                                                                                                   
tx-checksumming: off                                                                                                   
        tx-checksum-ipv4: off [fixed]                                                                                  
        tx-checksum-ip-generic: off                                                                                    
        tx-checksum-ipv6: off [fixed]                                                                                  
        tx-checksum-fcoe-crc: off [fixed]                                                                              
        tx-checksum-sctp: off [fixed]                                                                                  
scatter-gather: off                                                                                                    
        tx-scatter-gather: off                                                                                         
        tx-scatter-gather-fraglist: off [fixed]                                                                        
tcp-segmentation-offload: off                                                                                          
        tx-tcp-segmentation: off                                                                                       
        tx-tcp-ecn-segmentation: off [fixed]                                                                           
        tx-tcp-mangleid-segmentation: off                                                                              
        tx-tcp6-segmentation: off                                                                                      
generic-segmentation-offload: off                                                                                      
generic-receive-offload: off
large-receive-offload: off
rx-vlan-offload: off
tx-vlan-offload: off
ntuple-filters: on
receive-hashing: off [requested on]
highdma: on [fixed]
rx-vlan-filter: off                                                                                                    
vlan-challenged: off [fixed]                                                                                           
tx-lockless: off [fixed]                                                                                               
netns-local: off [fixed]                                                                                               
tx-gso-robust: off [fixed]                                                                                             
tx-fcoe-segmentation: off [fixed]                                                                                      
tx-gre-segmentation: off                                                                                               
tx-gre-csum-segmentation: off 
tx-ipxip4-segmentation: off
tx-ipxip6-segmentation: off
tx-udp_tnl-segmentation: off
tx-udp_tnl-csum-segmentation: off
tx-gso-partial: off
tx-tunnel-remcsum-segmentation: off [fixed]
tx-sctp-segmentation: off [fixed]
tx-esp-segmentation: off [fixed]
tx-udp-segmentation: off [requested on]
tx-gso-list: off [fixed]
fcoe-mtu: off [fixed]
tx-nocache-copy: off
loopback: off [fixed]
rx-fcs: off
rx-all: off
tx-vlan-stag-hw-insert: off
rx-vlan-stag-hw-parse: off [fixed]
rx-vlan-stag-filter: on [fixed]
l2-fwd-offload: off [fixed]
hw-tc-offload: off
esp-hw-offload: off [fixed]
esp-tx-csum-hw-offload: off [fixed]
rx-udp_tunnel-port-offload: off
tls-hw-tx-offload: off [fixed]
tls-hw-rx-offload: off [fixed]
rx-gro-hw: off [fixed]
tls-hw-record: off [fixed]
rx-gro-list: off
macsec-hw-offload: off [fixed]
```

**Comment by udippant** [comment]
On setting I see different from mine is `receive-hashing: on`. Not sure if that is necessarily the setting that enables it.

If that doesn't work I'll ask kernel experts internally about this.

**Comment by takehaya** [comment]
It seems that for some reason, the settings are not working. 
I don't know why it's off... 🤔 

If you know someone who is a kernel expert, ask them 🙇

**Comment by tehnerd** [comment]
@takehaya why do you think that rss is not working? do you have any data which shows this? just for the context. "working RSS - means that packets are evenly distributed amongs ingress NIC's queues". it is not some special magic option. yes some nics do understand IPIP and could look into inner packet. but for katran it is not prerequest. it is do mangling of outer packet (source address). in such scenario, if NIC does not understand protocol (for ipip it is 4 or 41). it tries to do RSS based on source and destination IP addresses.

tl;dr: please show data that RSS does not works. e.g. ```ethtool -S eth0 | grep rx``` where it is obvious that one of the queue receive more traffic than others

**Comment by takehaya** [comment]
@tehnerd Thanks for the comment.

I can't present the data right now. (I've shut down the remote server. sorry :bow: )
However, I can explain why I think RSS is not working.
The reason is that when I sent IPIP packets and looked at the CPU load in top/htop, only one cpu core was loaded.

To state a more sane example.
> watch -d -n 1 "grep mlx5 /proc/interrupts"

The value of IPIP does not change, but the value of IP, IP/UDP, and IP/TCP does.

**Comment by tehnerd** [comment]
what kind of traffic going to this server? is this some production or synthetic tests? how many flows?

**Comment by takehaya** [comment]
This is a multi-flow packet with different source addresses.

https://github.com/facebookincubator/katran/issues/112#issue-802044574
```
this case

v4 only: working
v4/udp: working
v4/tcp: working
v4/v4: not working (outer src-addr: different)

send packet
src-addr: different
dst-addr: same
```

**Comment by udippant** [comment]
Similar to what @tehnerd commented above, here's a summary of the responses I got internally: 

- There is no knob for IPIP RSS in kernel API. It is possible to add it in future but it has not been asked for.
-  NICs which support IP-IP RSS will hash on the inner headers. Sometime this behavior is controllable by vendor tools (or special FW builds).

So, either way, with or without the support for IPIP, you should be able to see RSS behavior on the RX path. 

If you can share the tcpdump on the receive side, it'll help us see what exact payload you're receiving, and see if that's inline with our expectation regarding the distribution of packets.

**Comment by tehnerd** [comment]
yeah. as well as the output of ethtool -S <intf_name>
